### PR TITLE
API: add numeric_only support to groupby agg

### PR DESCRIPTION
**Pull request by nicholas-ys-tan**
Currently in early phase seeking prelim feedback.

- [x] partially addresses #56946, relevant to #50538
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by nicholas-ys-tan** [comment]
Currently still not operating on the right numbers on test case `numeric_only True - 1_grouper_str - numpy_mean`.

Need to add test cases with expected data.

**Comment by nicholas-ys-tan** [comment]
> Currently still not operating on the right numbers on test case `numeric_only True - 1_grouper_str - numpy_mean`.
> 
> Need to add test cases with expected data.

Failing may be an existing bug - documented here: https://github.com/pandas-dev/pandas/issues/58146

File: pandas/core/base.py [hunk 1]
```diff
@@ -219,6 +219,25 @@ def _obj_with_exclusions(self):
         else:
             return self.obj
 
+    @final
+    @cache_readonly
+    def _obj_numeric_only_with_exclusions(self):
```
Comment (rhshadrach) [inline pandas/core/base.py:224]: I think the performance of this change needs to be closely studied to see if it is really worth the complexity (and extra memory usage). In any case, I do not think it should be introduced here, but rather in its own PR. Can you follow the existing pattern of subsetting the data directly in the `agg` methods instead.

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1605,19 +1608,19 @@ def _python_agg_general(self, func, *args, **kwargs):
...
 
-    def _aggregate_frame(self, func, *args, **kwargs) -> DataFrame:
+    def _aggregate_frame(self, func, *args, numeric_only=False, **kwargs) -> DataFrame:
         if self._grouper.nkeys != 1:
             raise AssertionError("Number of keys must be 1")
 
-        obj = self._obj_with_exclusions
-
+        mgr = self._get_data_to_aggregate(numeric_only=numeric_only)
+        data = self._wrap_agged_manager(mgr)
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:1616]: I don't see any advantage (or disadvantage - for that matter) of calling this obj vs data. Because of this, I think we should leave it as obj (to avoid the code churn).

If you see a reason to prefer data over obj, let me know!

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 3]
```diff
@@ -1663,3 +1663,53 @@ def func(x):
...
+        ["sum", np.mean],
+        lambda x: x.mean(),
+        {"A": "mean"},
+        {"A": "mean", "B": "sum"},
+        {"A": np.mean},
+    ],
+    ids=[
+        " string_mean ",
+        " numpy_mean ",
+        " list_of_str_and_str ",
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1684]: Why the spaces?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 4]
```diff
@@ -1663,3 +1663,53 @@ def func(x):
...
+        {
+            "A": [1, 2, 3, 4, 5],
+            "B": [10, 20, 30, 40, 50],
+            "groupby1": ["diamond", "diamond", "spade", "spade", "spade"],
+            "groupby2": [1, 1, 1, 2, 2],
+            "attr": ["a", "b", "c", "d", "e"],
+        }
+    )
+    if numeric_only or isinstance(aggfunc, dict):
+        df.groupby(by=groupers).agg(func=aggfunc, numeric_only=numeric_only)
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1712]: You need to check that the result is as expected. See other tests above for examples.

File: pandas/core/apply.py [hunk 5]
```diff
@@ -303,13 +303,14 @@ def agg_list_like(self) -> DataFrame | Series:
...
-        return self.agg_or_apply_list_like(op_name="agg")
+        kwargs = self.kwargs
+        return self.agg_or_apply_list_like(op_name="agg", **kwargs)
 
     def compute_list_like(
         self,
         op_name: Literal["agg", "apply"],
         selected_obj: Series | DataFrame,
-        kwargs: dict[str, Any],
+        **kwargs: dict[str, Any],
```
Comment (rhshadrach) [inline pandas/core/apply.py:313]: Can you revert this change. If you make it `**kwargs`, then users will run into an error if they try to pass through e.g. `selected_obj` as a keyword argument.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! It looks like you are using `**kwargs` to pass `numeric_only` in the call to `agg`. These `kwargs` are meant for the user to pass through to the function and I do not think we should be allowing "built-in" arguments into them. Instead, `numeric_only` should be added to the signature of `agg`.

~In addition, it appears that there are various paths through `agg` that do not respect `numeric_only`, e.g. `df.groupby(...).agg("sum", numeric_only=True)`.~ because `**kwargs` is passed through, this was wrong!

**Comment by nicholas-ys-tan** [comment]
> Thanks for the PR! It looks like you are using `**kwargs` to pass `numeric_only` in the call to `agg`. These `kwargs` are meant for the user to pass through to the function and I do not think we should be allowing "built-in" arguments into them. Instead, `numeric_only` should be added to the signature of `agg`.
> 
> ~In addition, it appears that there are various paths through `agg` that do not respect `numeric_only`, e.g. `df.groupby(...).agg("sum", numeric_only=True)`.~ because `**kwargs` is passed through, this was wrong!

Thank you for this note, I had assumed that we did not want to explicitly take `numeric_only` as a `kwarg` based on this comment:
https://github.com/pandas-dev/pandas/issues/49352#issuecomment-1304799900

If taking `numeric_only` as an arguement is the preferred method, this would probably greatly simplify things. One of my big struggles was the paths changing with and without `kwargs` but that can be more easily managed if we aren't feeding through `kwargs`.

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 4 above)
Comment (nicholas-ys-tan) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1712]: yup, sorry, I was in the middle of this ticket when I encountered the bug around #39169 which was an issue because `numeric_only` being fed as a `kwarg` sends it down this bugged path, then raised #58132 hoping to address that. We've discussed over there that this silent fix isn't ideal and we will be deprecating.

I will look into submitting a fix with an explicit `numeric_only` argument, as that can change the code path management.

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (nicholas-ys-tan) [reply pandas/core/groupby/generic.py:1616]: yeah, the issue I was having on this point was because `_obj_with_exclusions` doesn't update when the dataframe is updated to remove non-numeric columns. I assume this is related to it being a `@cache_readonly` in `pandas.core.base.py`. My tentative workaround was to generate the data again. 

On my local branch, I simplified this by creating another object:

```
@cache_readonly
def _obj_numeric_only_with_exclusions(self):
```

and in those segments:

```
if numeric_only:
    obj = self._obj_numeric_only_with_exclusions(self):
else:
    obj = self._obj_with_exclusions
```

Do you have any thoughts on this approach?

File: pandas/core/base.py (hunk 1 above)
Comment (nicholas-ys-tan) [reply pandas/core/base.py:224]: related to comment I put below:

https://github.com/pandas-dev/pandas/pull/58132#discussion_r1569203503

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 3 above)
Comment (nicholas-ys-tan) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1684]: Think it was ease for reading for myself when it was showing up on my tests, so it was easier to read

`xxx - yyyy - zzzz`

Will make sure to cleanup before taking off draft marker though.

**Comment by rhshadrach** [comment]
> Thank you for this note, I had assumed that we did not want to explicitly take `numeric_only` as a `kwarg` based on this comment:
> [#49352 (comment)](https://github.com/pandas-dev/pandas/issues/49352#issuecomment-1304799900)

That was if we were to fix it in a patch version (1.5.2). For 3.0, I think we should prefer a new argument.

> If taking `numeric_only` as an arguement is the preferred method, this would probably greatly simplify things.

It would also be more straightforward for users!

> Note that this was a half-cooked draft PR because I ran into #39169 bug

Yes - I think I failed to notice it was in draft status when I first reviewed. My bad.

**Comment by nicholas-ys-tan** [comment]
Closing as much of this will be addressed with a different PR, a fundamental misunderstanding of whether `numeric_only` should be its own kwarg, and the changes here being breaking changes that require more performance assessment.

Thank you to @rhshadrach for your time and your feedback on this PR. Will keep the above in mind for any future contributions I make.
